### PR TITLE
Asyncio

### PR DESCRIPTION
**Pull request by sphuber**
This is the rebased version of PR #150 . Originally the plan was to merge these changes in the `asyncio` branch before merging into develop because it will be backwards incompatible. However, since the last version `0.15.0` to support `tornado` has already been released we can proceed to merge directly into `develop`.

All the work was done by @unkcpz , I merely rebased it and prepared the commits.

### CODE REVIEW

File: test/test_processes.py [hunk 1]
```diff
@@ -557,28 +580,58 @@ def test_process_stack_multiple(self):
...
-        for _ in range(100):
+        n_run = 3
+        for _ in range(n_run):
             to_run.append(ParentProcess().step_until_terminated())
 
-        self.loop.run_sync(gen.coroutine(lambda: (yield gen.multi(to_run))))
+        loop = asyncio.get_event_loop()
+        loop.run_until_complete(asyncio.gather(*to_run))
+
+        for res in expect_true:
```
Comment (muhrin) [inline test/test_processes.py:612]: Could this be `assert all(expect_true)`

File: examples/process_launch.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        spec.inputs.dynamic = True
+        spec.outputs.dynamic = True
+        spec.output('default', valid_type=int)
+
+    def run(self):
+        self.out('default', 5)
+
+
+def main():
+    with rmq.RmqThreadCommunicator.connect(connection_params={'url': 'amqp://127.0.0.1'}
```
Comment (muhrin) [inline examples/process_launch.py:25]: Could this connect call be simplified to:

`kiwipy.connect('amqp://127.0.0.1')`

it just looks a little cleaner and I've changed the `kiwipy` docs to all use that too.

**Comment by merkys** [comment]
Glad to see this merged! Any plans to release soon?

**Comment by sphuber** [comment]
> Glad to see this merged! Any plans to release soon?

Absolutely, we are just looking if #147 and #161 need to be addressed first, but otherwise I will release `0.16.0` a.s.a.p. Note that we still need some work on `aiida-core`'s side to make that also fully independent of `tornado`. @unkcpz already opened [a PR](https://github.com/aiidateam/aiida-core/pull/4218) but it still needs some final touches and testing.

**Comment by sphuber** [comment]
@merkys I just released [v0.16.0](https://pypi.org/project/plumpy/)

**Comment by merkys** [comment]
> @merkys I just released [v0.16.0](https://pypi.org/project/plumpy/)

Great, thanks!

**Pull request merged**